### PR TITLE
Add guard-eval benchmarks for models with many repeated nn.Modules

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #188103
* #188101
* __->__ #187868

Dynamo cache-lookup / guard-evaluation cost scales with the number of
inlined nn.Modules: each repeated submodule contributes per-layer scalar
attribute guards, object-aliasing guards from shared objects, and
relational symbolic-shape guards. For deep stacks (e.g. many-layer decoder
models) this dominates the pre-graph runtime overhead, but no benchmark
exercised it -- existing nn.Module benchmarks measure compile time, and the
existing guard-eval microbenchmark uses a flat function rather than repeated
modules. See https://github.com/pytorch/pytorch/issues/185886.

This adds two benchmarks over a stack of repeated user-defined nn.Modules
carrying those guard patterns:

- A pr_time_benchmarks regression gate (nn_module_guard_eval.py) that measures
  runtime instruction count of guard_manager.check() in isolation, so the
  metric reflects guard-eval cost rather than graph execution. Deterministic
  instruction counts make it suitable for CI gating.
- A repeated_modules[N] case in the dynamo_guard_eval microbenchmark for quick
  local inspection of how guard eval scales with depth.

The expected_results.csv values are the instruction counts measured by the
trunk pr_time_benchmarks CI runner (nn_module_guard_eval 80120,
nn_module_guard_eval_dynamic 111714).

Test Plan:

pr_time benchmark passes against the recorded baseline:

```
cd benchmarks/dynamo/pr_time_benchmarks
PYTHONPATH=./benchmarks python benchmarks/nn_module_guard_eval.py result.csv
PYTHONPATH=./benchmarks python check_results.py expected_results.csv result.csv new.csv
# -> All benchmarks passed
```

Microbenchmark runs and shows the expected linear scaling:

```
python benchmarks/dynamo/microbenchmarks/dynamo_guard_eval.py
# repeated_modules[10] 11.9us
# repeated_modules[40] 34.3us
# repeated_modules[80] 69.8us
```

This PR was authored with the assistance of an AI coding assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: benchmarks/dynamo/pr_time_benchmarks/expected_results.csv [hunk 1]
```diff
@@ -119,3 +119,11 @@ dtensor_dispatch_random,instruction_count,592800,0.1
 
 
 dtensor_dispatch_custom_handler,instruction_count,34910,0.1
+
+
+
+nn_module_guard_eval,instruction_count,82620,0.1
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/expected_results.csv]: [blocker] These baselines (82620 / 98899) were measured locally, per the PR description. `benchmark_runner.sh` auto-discovers every `benchmarks/*.py`, so `nn_module_guard_eval.py` runs unconditionally, and `.ci/pytorch/test.sh`'s `pr_time_benchmarks()` then runs `check_results.py`, which `sys.exit(1)`s on any entry whose CI value deviates more than the +-10% margin from this file. That gate lives in `trunk.yml` (config `pr_time_benchmarks`), not the default PR `pull` workflow -- so it will not fail a check on this PR, but it runs post-merge on `main` (or under `ciflow/trunk`) and will turn the trunk job red for everyone if the runner's instruction count lands outside +-10% of these local numbers. Safer sequencing: push `ciflow/trunk` on this PR, copy the real count from the job's "To update expected results" block into this file, then land -- rather than land red and fix forward.

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/nn_module_guard_eval.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            prefix += "_dynamic"
+        return prefix
+
+    def description(self):
+        return "runtime guard-eval cost for a stack of repeated nn.Modules"
+
+    def _prepare_once(self):
+        torch._dynamo.reset()
+        self._model = RepeatedModules(self.N_LAYERS, self.HIDDEN)
+        compiled = torch.compile(
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/nn_module_guard_eval.py]: [cleanup] The `instruction_count` path (`BenchmarkBase._count_instructions`) does not pin `enable_cpp_symbolic_shape_guards`; only the `compile_time_instruction_count` path patches it to False in `collect_all()`, and benchmark_base's own comment notes that flag "has impact on these benchmarks." Because `torch._C._instruction_counter` counts the C++ instructions executed under `guard_manager.check()`, the `nn_module_guard_eval_dynamic` baseline here is measured against whatever the ambient default is (currently False). If that default ever flips to True, the symbolic-shape guards move into a compiled C++ guard and this benchmark's instruction count shifts substantially with no change to this file -- silently breaking the gate. Pinning it explicitly around the compile (e.g. `with torch._dynamo.config.patch(enable_cpp_symbolic_shape_guards=False):`) makes the dynamic baseline robust and matches the compile-time benchmarks. The static variant has no symbolic-shape guards and is unaffected.

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/nn_module_guard_eval.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+class GuardEvalLayer(nn.Module):
+    def __init__(self, idx, hidden, shared):
+        super().__init__()
+        self.eps = 1e-6
+        self.scale = 0.5 + idx * 1e-3
+        self.register_buffer("bias", torch.randn(hidden))
+        self.shared = shared
+
+    def forward(self, x):
+        if self.shared.flag:
```
Comment (aorenste) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/nn_module_guard_eval.py:36]: [cleanup] This read is load-bearing but looks like dead control flow: `flag` is always True, so the `if` is constant-folded and `x = x + self.eps` is unconditional. Its real purpose is to dereference the shared object during tracing, which is what makes Dynamo emit the object-aliasing guards (`layers[i].shared is layers[j].shared`) that this benchmark is meant to exercise. Verified empirically: remove this read and the N-1 aliasing guards vanish entirely, while the per-layer scalar/tensor/shape guards stay. Worth a one-line comment marking it load-bearing, so a future cleanup doesn't "simplify" the always-true branch away and silently gut the benchmark's headline guard pattern. (Same applies to `_RepeatedLayer.forward` in the `dynamo_guard_eval.py` microbenchmark.)

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

Approving. The inline comments should be addressed before landing -- in particular the [blocker] on `expected_results.csv`: those baselines are local measurements, and the trunk `pr_time_benchmarks` gate will go red post-merge if the CI instruction count lands outside the +-10% margin. Recommend pushing `ciflow/trunk` to capture a real CI number, committing it, then landing.

[cleanup] The `Shared`/`GuardEvalLayer`/`RepeatedModules` classes here duplicate `_Shared`/`_RepeatedLayer`/`_RepeatedModules` in `benchmarks/dynamo/microbenchmarks/dynamo_guard_eval.py` almost verbatim. The two benchmark suites can't easily share a module, so some duplication is unavoidable, but both copies model the exact same guard patterns and will silently drift apart if one is later updated and the other isn't. A short cross-reference comment in each file pointing at the other would keep a future edit honest.
